### PR TITLE
US_21 

### LINKED ISSUE #26: 21. Admin Dashboard Statistics - Top Customers
As an admin,
When I visit the admin dashboard (/admin)
Then I see the names of the top 5 customers
who have conducted the largest number of successful transactions
And next to each customer name I see the number of successful transactions they have
conducted

### PR DESCRIPTION
**Pull request by Daniel-Gallagher92**
__3__ Wrote Tests ____ Implemented ____ Reviewed

Neccesary checkmarks:

    [x] All Tests are Passing

    [x] The code will run locally

Type of change

    [x] New feature
    [x] Bug Fix

Implements/Fixes:

    description closes #

Check the correct boxes

    [x] This broke nothing
    [] This broke some stuff
    [] This broke everything

Testing Changes

    [] No Tests have been changed
    [] Some Tests have been changed
    [] All of the Tests have been changed(Please describe what in the world happened)

Checklist:

    [x] My code has no unused/commented out code
    [x] I have reviewed my code
    [] I have commented my code, particularly in hard-to-understand areas
    [x] I have fully tested my code

(REQUIRED)Please Include a link to a gif of your feelings about this branch

Link:
https://media.giphy.com/media/xT5LMHxhOfscxPfIfm/giphy.gif

### CODE REVIEW

**Review by JorjaF — COMMENTED** [review]
I think this looks really good to me

File: spec/features/admin/dashboard_spec.rb [hunk 1]
```diff
@@ -1,25 +1,145 @@
...
+    visit "/admin"
+    save_and_open_page
+    within ".top_customers" do 
+      expect("Heber").to appear_before("Cecelia")
+      expect("Cecelia").to appear_before("Mariah")
+      expect("Mariah").to appear_before("Sylvester")
+      expect("Sylvester").to appear_before("Joey")
+      expect("Joey").to_not appear_before("Heber")
+      expect(page).to_not have_content("Leanna")
+
```
Comment (ethanrossblack) [inline spec/features/admin/dashboard_spec.rb:153]:
```suggestion
    end
```

Comment (ethanrossblack) [reply spec/features/admin/dashboard_spec.rb:153]: close out the within block

File: spec/models/customer_spec.rb [hunk 2]
```diff
@@ -1,5 +1,140 @@
...
+        it 'displays the names of the top 5 customers by successful transactions' do 
+          expect(Customer.top_5_customers).to match([@customer_6, @customer_2, @customer_3, @customer_5, @customer_1])
+          expect(Customer.top_5_customers[0].first_name).to match("Heber")
+        end
+      end
+
+      describe ".most_successful_transactions" do
+        it "displays the transaction count of the top 5 customers" do
+          expect(@customer_3.successful_transaction_count).to eq(4)
+          expect(@customer_2.successful_transaction_count).to eq(5)
```
Comment (z-fitch) [inline spec/models/customer_spec.rb:137]: Very nice tests!

File: spec/models/customer_spec.rb [hunk 3]
```diff
@@ -1,5 +1,140 @@
...
+  end
+
+  describe "Admin user story set up" do
+
+      #US_21
+
+    describe "class methods" do 
+      describe ".top_customers" do 
+        it 'displays the names of the top 5 customers by successful transactions' do 
+          expect(Customer.top_5_customers).to match([@customer_6, @customer_2, @customer_3, @customer_5, @customer_1])
```
Comment (ethanrossblack) [inline spec/models/customer_spec.rb:129]: What does `.match` do? Would `.match_array` work better here?

(For more reading: https://medium.com/beambenefits/array-matchers-in-rspec-42b370e3983e)

File: spec/models/customer_spec.rb [hunk 4]
```diff
@@ -1,5 +1,140 @@
...
+    it {should have_many(:merchants).through(:items)}
+  end
+
+  describe "Admin user story set up" do
+
+      #US_21
+
+    describe "class methods" do 
+      describe ".top_customers" do 
+        it 'displays the names of the top 5 customers by successful transactions' do
```
Comment (ethanrossblack) [inline spec/models/customer_spec.rb:128]: doesn't it `display the names` of the top 5 customers etc. or does it collect them in an array?

File: spec/features/admin/dashboard_spec.rb [hunk 5]
```diff
@@ -1,25 +1,145 @@
...
       expect(page).to have_link("Invoices")
   end
-# US 22  
+
+
+  #US 21
+  it "displays a list of the top 5 customers" do 
+
+    visit "/admin"
+    save_and_open_page
```
Comment (z-fitch) [inline spec/features/admin/dashboard_spec.rb]: Doesn't need to be done now, but should prob get rid of the save_and_open_page

**Review by z-fitch — APPROVED** [review]

File: spec/models/customer_spec.rb [hunk 6]
```diff
@@ -1,5 +1,140 @@
...
+
+  describe "Admin user story set up" do
+
+      #US_21
+
+    describe "class methods" do 
+      describe ".top_customers" do 
+        it 'displays the names of the top 5 customers by successful transactions' do 
+          expect(Customer.top_5_customers).to match([@customer_6, @customer_2, @customer_3, @customer_5, @customer_1])
+          expect(Customer.top_5_customers[0].first_name).to match("Heber")
```
Comment (ethanrossblack) [inline spec/models/customer_spec.rb:130]: Zanna and I wrote a class method for Customer called `.full_name` that stringifies a customer's first+last name, if you wanted to include that here (or after we merge our PR)

File: spec/models/customer_spec.rb [hunk 7]
```diff
@@ -1,5 +1,140 @@
...
+        end
+      end
+
+      describe ".most_successful_transactions" do
+        it "displays the transaction count of the top 5 customers" do
```
Comment (ethanrossblack) [inline spec/models/customer_spec.rb:135]: does it `display the transactions` or does it `return` that data?

File: spec/models/customer_spec.rb [hunk 8]
```diff
@@ -1,5 +1,140 @@
...
+    @transaction_30 = @invoice_10.transactions.create!(credit_card_number: '4923661117104166', credit_card_expiration_date: '08/22/20', result: 'success')
+
+    @transaction_31 = @invoice_11.transactions.create!(credit_card_number: '4923661117104166', credit_card_expiration_date: '08/22/20', result: 'success')
+    @transaction_32 = @invoice_11.transactions.create!(credit_card_number: '4923661117104166', credit_card_expiration_date: '08/22/20', result: 'success')
+    @transaction_33 = @invoice_11.transactions.create!(credit_card_number: '4923661117104166', credit_card_expiration_date: '08/22/20', result: 'success')
+
+    @transaction_34 = @invoice_12.transactions.create!(credit_card_number: '4923661117104166', credit_card_expiration_date: '08/22/20', result: 'success')
+    @transaction_35 = @invoice_12.transactions.create!(credit_card_number: '4923661117104166', credit_card_expiration_date: '08/22/20', result: 'success')
+    @transaction_36 = @invoice_12.transactions.create!(credit_card_number: '4923661117104166', credit_card_expiration_date: '08/22/20', result: 'success')
+  end
```
Comment (ethanrossblack) [inline spec/models/customer_spec.rb:112]: EXTREMELY thorough testing data. TY for this, we may have to use it for the Merchants data page. Would it be possible to summarize the test data in a comment? It's extremely dense rn and having a centralized place to quickly know what's inside, what should be expected would be helpful for future testing!

File: app/views/admin/index.html.erb [hunk 9]
```diff
@@ -10,6 +10,18 @@
     </nav>
 </div>
 
+
+<div class="column">
+  <section class="top_customers">
+    <h3>  Top 5 Customers</h3>
+      <ol>
+        <% @top_5_customers.each do |customer| %>
+          <li><%="#{customer.first_name} #{customer.last_name} has made #{customer.result_count} successful transactions"%></li><br>
```
Comment (ethanrossblack) [inline app/views/admin/index.html.erb:19]: potential refactor using the `customer.full_name` instance method Zanna and I put together

Comment (ethanrossblack) [reply app/views/admin/index.html.erb:19]: Also, what is `customer.result_count`? I don't see that as an instance method anywhere.

File: spec/features/admin/dashboard_spec.rb (hunk 5 above)
Comment (Daniel-Gallagher92) [reply spec/features/admin/dashboard_spec.rb]: Mod 1 - leaves in pry
Mod 2 - leaves in saop 
😅

**Review by ethanrossblack — CHANGES_REQUESTED** [review]
Looks good! See my comments. Great work!

File: spec/features/admin/dashboard_spec.rb (hunk 5 above)
Comment (z-fitch) [reply spec/features/admin/dashboard_spec.rb]: its a rite of passage

**Review by ethanrossblack — APPROVED** [review]
good stuff!

**Pull request merged**